### PR TITLE
Add an introduction to tracing lab

### LINKED ISSUE #3: Why/when tracing
Do we need this, what will be shown?

We need to have an example problem?

Not findable with debug or with profilers.

Link tracing with monitoring?
Link with logs?

### PR DESCRIPTION
**Pull request by tahini**
Fixes #3

Signed-off-by: Geneviève Bastien <gbastien+lttng@versatic.net>

### CODE REVIEW

File: labs/00-what-is-tracing/README.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
+### What /when/why tracing
```
Comment (TheMatthew) [inline labs/00-what-is-tracing/README.md]: Titles should have every first letter capitalized in english.

File: labs/00-what-is-tracing/README.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+In for loop 2                          // <-- event
+Done: 10                               // <-- event
+```
+
+But printf is usually used for quick instantaneous debugging of an application and is removed from production code (right?)
+
+Logging is another form of tracing. It usually associates a timestamp with events to better understand the timing in the application. But we usually log only high level information, as disk space is a limited resource (right?)
+
+The `tracing` we're discussing here is high speed, low overhead tracing. With such tracing, the tracepoints can be present in the code at all time (linux has tons of tracepoints in its code, ready to be hooked to), they have a near-zero overhead when not tracing and a very low one one with a tracer enabled. Tracers can store hundreds of thousands events/second. Try that with printf!!
+
```
Comment (TheMatthew) [inline labs/00-what-is-tracing/README.md]: What about d-trace like aggregators?

File: labs/00-what-is-tracing/README.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+When these tools fails though, `tracing` is there:
+
+* When `profiling` will miss that call at around 17min in the execution where there was a huge latency, it will either go unnoticed, or be lost as a statistic, without context or detail. `Tracing` can detect that latency and you can then zoom in on it and see everything that happened around that time and led to that moment.
+
+* When `debugging` your algorithm, or even printing the details, the bug never happen. The added latency of those techniques makes that race condition impossible to catch! Or that unit test that fails one in a 100 times. `Tracing`'s overhead is rarely high enough to remove the potential race condition, so if running with tracing enabled, when the problem happens, the trace may show what happened on the system at the time, or you may need to enable more events, but it will eventually show something.
+
+* When `log` files have put you on a trail, but looking at everything, you just can't put your finger on it. `Tracing` will show everything (well, almost) that is hidden from log data and you can have an overview of the whole system at a given time.
+
```
Comment (TheMatthew) [inline labs/00-what-is-tracing/README.md]: When timing will affect the program. E.G. the server will timeout when you have a breakpoint

File: labs/00-what-is-tracing/README.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+Before going further in this tutorial and starting with the labs, let's define tracing: what it is, what it is not, when to use it and how to compare it with other tools.
+
+- - -
+
+#### What is tracing?
+
+`Tracing` consists in recording specific information during a program's or operating system's execution to better understand what is happening on the system. The simplest form of tracing is what we all learn in programming 101: printf!
+
+Every location in the code that we want to trace is called a `tracepoint` and every time a tracepoint is hit, it creates an `event`. The following example shows those concepts
```
Comment (TheMatthew) [inline labs/00-what-is-tracing/README.md]: In aggregators it can increment a timer or performa an action

**Comment by TheMatthew** [comment]
Also, what about sampling?

**Comment by tahini** [comment]
I think I addressed all of your comments with some rephrasing here and there

File: labs/00-what-is-tracing/README.md (hunk 2 above)
Comment (tuxology) [reply labs/00-what-is-tracing/README.md]: I think we should differentiate between trace aggregation and trace collection properly. Trace collectors are like `perf trace` LTTng etc while aggregators are like DTrace, `bpftrace` (which uses eBPF)

Comment (tahini) [reply labs/00-what-is-tracing/README.md]: tuxology look at the new version of this paragraph. and see if it's ok wrt that

**Comment by TheMatthew** [comment]
I am OK with merging this as-is and amending if needed

**Comment by tahini** [comment]
Ok let's merge then

**Pull request merged**